src/operators/join.rs

```diff
@@ -484,24 +507,43 @@ impl<G, T1> JoinCore<G, T1::Key, T1::Val, T1::R> for Arranged<G,T1>
                     activator.activate();
                 }
 
-                // shut down or advance trace2.
-                if trace2.is_some() && input1.frontier().is_empty() { trace2 = None; }
-                if let Some(ref mut trace2) = trace2 {
-                    trace2.set_logical_compaction(input1.frontier().frontier());
-                    trace2.advance_upper(&mut acknowledged2);
-                    // We validated that `ack2` was at least `distinguish_since()` at start-up.
-                    // It has only advanced since then, so this should be safe to do.
-                    trace2.set_physical_compaction(acknowledged2.borrow());
+                // 3. Trace maintenance.
+                //
+                // Importantly, we use `input.frontier()` here rather than `acknowledged` to track
+                // the progress of an input, because should we ever drop one of the traces we will
+                // lose the ability to extract information from anything other than the input.
+                // For example, if we dropped `trace2` we would not be able to use `advance_upper`
+                // to keep `acknowledged2` up to date wrt empty batches, and would hold back logical
+                // compaction of `trace1`.
+
+                // Maintain `trace1`. Drop if `input2` is empty. or advance based on future needs.
+                if let Some(trace1) = trace1_option.as_mut() {
+                    if input2.frontier().is_empty() { trace1_option = None; }
+                    else {
+                        // Allow `trace1` to logically compact up to the frontier we may yet receive,
+                        // in the opposing input (`input2`). All `input2` times will be beyond this
+                        // frontier, and joined times only need to be accurate when advanced to it.
+                        trace1.set_logical_compaction(input2.frontier().frontier());
+                        // Allow `trace1` to physically compact up to the upper bound of batches we
+                        // have received in its input (`input1`). We will not require a cursor that
+                        // is not beyond this bound.
+                        trace1.set_physical_compaction(acknowledged1.borrow());
```

Comment:
does acknowledged1 need to be <= input2.frontier() / is it always because of some other invariant?

Reply:
No, they are generally unrelated.